zerver/lib/message.py

```diff
@@ -408,13 +429,25 @@ def has_message_access(
         # You can't access messages in deactivated streams
         return False
 
-    def is_subscribed_helper() -> bool:
-        if is_subscribed is not None:
-            return is_subscribed
+    def has_content_access_helper() -> bool:
+        if is_subscribed is None:
+            is_user_subscribed = Subscription.objects.filter(
+                user_profile=user_profile, active=True, recipient=message.recipient
+            ).exists()
+        else:
+            is_user_subscribed = is_subscribed
+
+        if is_user_subscribed:
+            return True
 
-        return Subscription.objects.filter(
-            user_profile=user_profile, active=True, recipient=message.recipient
-        ).exists()
+        if user_group_membership_details.user_recursive_group_ids is None:
+            user_group_membership_details.user_recursive_group_ids = set(
+                get_recursive_membership_groups(user_profile).values_list("id", flat=True)
+            )
+
+        return is_user_in_groups_granting_content_access(
+            stream, user_group_membership_details.user_recursive_group_ids
```

Comment:
The `Stream` fetched above, via `Stream.objects.get(id=message.recipient.type_id)`, likely is missing a `selected_related`. Probably some direct tests with query counts for corner cases like this would be helpful.

Reply:
We don't need to fetch related groups if that is what you were mentioning the select_related to be used. We're just using the ids directly again user_recursive_group_ids, I am replacing `stream.recipient` with `stream.recipient_id` though.

Reply:
Yeah, that makes sense, I misread. We'd need it only if we're doing `NamedUserGroup` stuff. Thanks for correcting!